onnxscript/ir/_external_data.py

```diff
@@ -51,3 +75,209 @@ def set_base_dir(graph: _core.Graph | _core.GraphView, base_dir: str | os.PathLi
+) -> int:
+    """
+    Method to compute updated offset.
+
+    Args:
+        current_offset: Current location in the file at which tensor data will be written to.
+        tensor_size: Size of the tensor data to be written to file.
+        align_offset: Offset will always be page aligned and alloction granularity aligned for mmap support. This is done by padding previous tensor data with zeros keeping same length. Tensor data will be aligned if > align_threshold
+        align_threshold: Alignment threshold for size of data. Having a low threshold will waste file space for small initializers. Only when tensor's data is > the page_align_threshold it will be force aligned.
+        allocation_granularity: The allocation Granularity for mmap() support. Typically 64KB for Windows & 4KB for other OSes.
+    """
+    if align_offset and tensor_size > align_threshold:
+        alignment_factor = max(4096, allocation_granularity)
+        # Align to the next page or alloc granularity
+        current_offset = (
+            (current_offset + alignment_factor - 1) // alignment_factor * alignment_factor
+        )
+    return current_offset
+
+
+def _set_external_data(
+    tensor: _protocols.TensorProtocol,
+    current_offset: int,
+) -> ExternalDataInfo:
+    """Method to capture information about a tensor that is to be stored as external data."""
+    tensor_size = tensor.nbytes
+    # Calculate updated offset and align tensors
+    current_offset = compute_new_offset(current_offset, tensor_size)
+    # Store offset and tensor size as ExternalDataInfo
+    external_data_info = ExternalDataInfo(
+        tensor.name,  # type: ignore[arg-type]
+        current_offset,
+        tensor_size,
+    )
+    return external_data_info
+
+
+def _save_external_data(
+    external_data_info: list[tuple[_protocols.TensorProtocol, ExternalDataInfo]],
+    file_path: str | os.PathLike,
+) -> None:
+    """
+    Writes tensor data to an external file according to information stored in ExternalDataInfo object.
+
+    Args:
+        external_data_info: A collection of external data information stored for each tensor to be written as external data.
+        file_path: Location to which external data is to be stored.
+    """
+    with open(file_path, "w+b") as data_file:
+        for tensor, tensor_info in external_data_info:
+            current_offset = tensor_info.offset
+            assert tensor is not None
+            raw_data = tensor.tobytes()
+            if current_offset is not None:
+                # Pad file to required offset if needed
+                file_size = data_file.tell()
+                if current_offset > file_size:
+                    data_file.write(b"\0" * (current_offset - file_size))
+            data_file.write(raw_data)
+
+
+def _convert_as_external_tensors(
+    external_data_info: list[tuple[_protocols.TensorProtocol, ExternalDataInfo]],
+    file_path: str | os.PathLike,
```

Comment:
Should we provide both the rel path and the base path? Otherwise the external tensor data cannot be accessed in the ir